### ISSUE TITLE
EXCEPTION_ACCESS_VIOLATION When using the same frame in different thread

### ISSUE THREAD

**Issue by k3rnL** [bytedeco/javacv#1161]
Hello,

I have some issue with using a Frame in three different thread (preview player, recorder, streamer). I got an EXCEPTION_ACCESS_VIOLATION when I convert it using Java2DFrameConverter and/or recording it using two different FFMpegFrameRecorder.

I'm pretty sure that is a concurrent access that give me the problem, even if I'm surprise that there is a problem with it, all those tasks should do only reading of the memory segment, right ? (maybe memery is freed up before ffmpeg use it?) 
To try to solve the problem I added some synchronize {} block around each of those tasks, and it gave me no exception. But in this case, it's useless to have 3 different thread, and it gave me a really low framerate (for example youtube show me a stream problem with 1000kbit/s).

I tried to clone each frame for their respective thread, but I got the same result and also my RAM just goes crazy.

Do you have any recommendation on how to find why and where an access violation happen ? 
Do you think that it's a mistake to have different thread for processing each frame ?

There is my access violation error :
```
[thread 8252 also had an error]# A fatal error has been detected by the Java Runtime Environment:

#
#  EXCEPTION_ACCESS_VIOLATION (0xc0000005) at pc=0x00007ff9265476d9, pid=11960, tid=14504
#
# JRE version: Java(TM) SE Runtime Environment (10.0.2+13) (build 10.0.2+13)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (10.0.2+13, mixed mode, tiered, compressed oops, g1 gc, windows-amd64)
# Problematic frame:
# C  [swscale-5.dll+0x576d9]
#
# No core dump will be written. Minidumps are not enabled by default on client versions of Windows
#
# An error report file with more information is saved as:
# C:\Users\Erwan\Documents\Workspace\Freelancer\srajca\test\LDSLiveSubtitles\hs_err_pid11960.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

And where it happens (I don't know which part is problematic) : 
The player :
```java
super.paintComponent(g);
        g.setColor(Color.BLACK);
        g.drawRect(getX(), getY(),
                getWidth(), getHeight());
        if (mActualImageFrame != null) {
            BufferedImage img;
//            synchronized (VideoReader.class) {
                img = new Java2DFrameConverter().convert(mActualImageFrame);
//            }
            g.drawImage(img,
                    mVideoOffset.width, mVideoOffset.height,
                    mVideoDimension.width, mVideoDimension.height, this);
        }
```
The streamer :
```java
private void stream() {
        Frame f = mFrameBuffer.poll();
        if (f == null)
            return;
        try {
//            synchronized (VideoReader.class) {
                mRecorder.record(f);
//            }
        } catch (FrameRecorder.Exception e) {
            e.printStackTrace();
        }
    }
```
The recorder :
```java
    private void record() {
        Frame f = mFrameBuffer.poll();
        if (f == null) {
   return;
        }
        try {
//            synchronized (VideoReader.class) {
                mRecorder.record(f);
//            }
        } catch (FrameRecorder.Exception e) {
            e.printStackTrace();
        }
    }
```

**Comment by saudet** [comment]
There has been changes since the last release that might fix that. Please
give it a try with the snapshots: http://bytedeco.org/builds/

**Comment by saudet** [comment]
If VideoReader uses FFmpegFrameGrabber, you'll need to make sure grab() doesn't get called. You could also copy the frame after grab() with Frame.clone() and use that one from other threads, but this is less efficient.

**Comment by saudet** [comment]
We could also enhance FFmpegFrameGrabber to relinquish ownership of each Frame it returns, and you would have to add them back manually. If this is something you need, please feel free to make a contribution it that sense and send a pull request!

**Comment by k3rnL** [comment]
Why "add them back manually" ?

I'm trying to use the snapshot, but I have some problem, I let you understand by yourself :
```java
Exception in thread "AWT-EventQueue-0" java.lang.NoClassDefFoundError: java.lang.ClassNotFoundException: org.bytedeco.ffmpeg.avutil.org.bytedeco.ffmpeg.global.avutil
	at org.bytedeco.javacpp.Loader.load(Loader.java:999)
	at org.bytedeco.javacpp.Loader.load(Loader.java:935)
	at org.bytedeco.ffmpeg.avformat.Read_packet_Pointer_BytePointer_int.<clinit>(Read_packet_Pointer_BytePointer_int.java:44)
	at org.bytedeco.javacv.FFmpegFrameGrabber.<clinit>(FFmpegFrameGrabber.java:310)
```

To take the snapshot I just changed the version number and added repository to my list, I think there is nothing wrong:
```xml
<dependency>
            <groupId>org.bytedeco</groupId>
            <artifactId>javacv</artifactId>
            <version>1.5-SNAPSHOT</version>
</dependency>
```

**Comment by saudet** [comment]
The package names changed a bit, use org.bytedeco.ffmpeg.

**Comment by saudet** [comment]
Also, we still should use javacv-platform.

**Comment by k3rnL** [comment]
Mmmh.. It seems that I still have the same error in swscale, but less often. It happens randomly and sometimes after few hours or few seconds.
After a lot of tries, I discovered that it's the use of java2dframeconverter in parallel with a ffmpegframerecorder that is problematic. 

I tried to take a look of the source code to try to understand why. I'm pretty sure that the frame converter change the position of the buffer during the creation of the pointer in the recorder, it makes a pointer with a wrong position and makes the native code crash by going further than it should (FFmpegFrameRecorder.java:904).
The frame converter doesn't care because on each iteration during the copy it reset the position on the next pixel (Java2DFrameConverter.java:114).
It's really a not lucky case, but it can happen. 

It also explain why I don't have crash when I use two recorder in the same time, they are always moving the position of the buffer on 0.

You can easily reproduce the bug by using the converter and the recorder in different thread and give them the same buffer in loop (Frame.image and Frame.opaque). It crashes instantly.

**Comment by saudet** [comment]
I see, we shouldn't rely on the position when accessing buffers, good
point. Let's what we can do.

**Comment by k3rnL** [comment]
A quick fix would be to change the code of the frame converter by using index, but it would be better to fix frame recorder. Can't we use the Frame.opaque pointer directly more than the Buffer in parameter of recordImage ?

**Comment by k3rnL** [comment]
Finally I changed both files. They does not change the position of the buffer anymore.

**Comment by saudet** [comment]
The changes have been released as part of JavaCV 1.5. Thanks for the contribution!
